src/lossless_transform.rs

```diff
@@ -402,59 +406,166 @@ pub(crate) fn apply_color_indexing_transform(
+    let expanded_lookup_table_storage: Vec<[u8; EXP_ENTRY_SIZE]> = (0..256u16)
+        .map(|packed_byte_value_u16| {
+            let mut entry_pixels_array = [0u8; EXP_ENTRY_SIZE]; // Uses const generic
+            let packed_byte_value = packed_byte_value_u16 as u8;
+
+            // Loop bound is effectively constant for each instantiation.
+            for pixel_sub_index in 0..pixels_per_packed_byte_usize {
+                let shift_amount = (pixel_sub_index as u8) * bits_per_entry_u8;
+                let k = (packed_byte_value >> shift_amount) & mask_u8;
+
+                let color_source_array: [u8; 4] = if k < table_size {
+                    let color_data_offset = usize::from(k) * 4;
+                    table_data[color_data_offset..color_data_offset + 4]
+                        .try_into()
+                        .unwrap()
                 } else {
-                    image_data[output_index..][..entry_size].copy_from_slice(table[table_index]);
-                }
+                    [0u8; 4] // WebP spec: out-of-bounds indices are [0,0,0,0]
+                };
+
+                let array_fill_offset = pixel_sub_index * 4;
+                entry_pixels_array[array_fill_offset..array_fill_offset + 4]
+                    .copy_from_slice(&color_source_array);
             }
+            entry_pixels_array
+        })
+        .collect();
+
+    let expanded_lookup_table_array: &[[u8; EXP_ENTRY_SIZE]; 256] =
+        expanded_lookup_table_storage.as_slice().try_into().unwrap();
+
+    let packed_image_width_in_blocks = width.div_ceil(pixels_per_packed_byte_u8.into()) as usize;
+
+    if width == 0 || height == 0 {
+        return;
+    }
+
+    let final_block_expanded_size_bytes =
+        (width as usize * 4) - EXP_ENTRY_SIZE * (packed_image_width_in_blocks.saturating_sub(1));
+
+    let input_stride_bytes_packed = packed_image_width_in_blocks * 4;
+    let output_stride_bytes_expanded = width as usize * 4;
+
+    let mut packed_indices_for_row: Vec<u8> = vec![0; packed_image_width_in_blocks];
+
+    for y_rev_idx in 0..height as usize {
+        let y = height as usize - 1 - y_rev_idx;
+
+        let packed_row_input_global_offset = y * input_stride_bytes_packed;
+        let packed_argb_row_slice =
+            &image_data[packed_row_input_global_offset..][..input_stride_bytes_packed];
+
+        for (packed_argb_chunk, packed_idx) in packed_argb_row_slice
+            .chunks_exact(4)
+            .zip(packed_indices_for_row.iter_mut())
+        {
+            *packed_idx = packed_argb_chunk[1];
+        }
+
+        let output_row_global_offset = y * output_stride_bytes_expanded;
+        let output_row_slice_mut =
+            &mut image_data[output_row_global_offset..][..output_stride_bytes_expanded];
+
+        let num_full_blocks = if packed_image_width_in_blocks > 0 {
+            packed_image_width_in_blocks - 1
+        } else {
+            0
+        };
```

Comment:
```suggestion
        let num_full_blocks = packed_image_width_in_blocks.saturating_sub(1);
```
Surprisingly reduces 400 lines of code size. I guess some of the paths are allowed to be reused.
https://rust.godbolt.org/z/rqzW7446a - Line 153

---

I was about to recommend the same change as https://github.com/image-rs/image-webp/pull/134/commits/fdb4ca812008d38e5e82d616c891e05f9995d0ba since it wasn't in the originally linked godbolt. The autovectorization code appears much better.